### PR TITLE
Fix for deadlock on grad sampling

### PR DESCRIPTION
**Pull request by eldarkurtic**
It fixes the deadlock by properly collecting gradients when model is wrapped with DDP. Each process gets it's own grad_sampler instance, they all perform forward-backward but only the main process gets to collect the gradients.

This enables reducing the strain on the GPU-0 when collecting gradients with large `grad_sampler_batch_size` by allowing us to reduce the `grad_sampler_batch_size` by a factor of `world_size` on each device and still get the same overall batch-size for gradients.

### CODE REVIEW

File: src/sparseml/pytorch/sparsification/pruning/modifier_pruning_obs.py [hunk 1]
```diff
@@ -293,28 +291,24 @@ def _collect_grad_samples(
...
-            module = module.module
         if not isinstance(grad_sampler, GradSampler):
             raise ValueError(
                 "One-shot OBS pruning requires a GradSampler object given by the "
                 f"grad_sampler kwarg. Given an object of type {type(grad_sampler)}"
             )
 
         is_training = module.training
-        _LOGGER.debug("Setting the model in the eval mode")
+        _LOGGER.info("Setting the model in the eval mode")
```
Comment (Godofnothing) [inline src/sparseml/pytorch/sparsification/pruning/modifier_pruning_obs.py:301]: Probably it's better to keep this message only in the DEBUG mode

File: src/sparseml/pytorch/sparsification/pruning/modifier_pruning_obs.py [hunk 2]
```diff
@@ -240,14 +238,12 @@ def initialize(
...
-        if self._scorer._is_main_proc:  # grads collected only in the main proc
-            self._grad_sampler = GradSampler(
-                kwargs["grad_sampler"]["data_loader_builder"](
-                    **self._grad_sampler_kwargs
-                ),
-                kwargs["grad_sampler"]["loss_function"],
-            )
+        self._grad_sampler = GradSampler(
+            kwargs["grad_sampler"]["data_loader_builder"](
+                self._grad_sampler_kwargs
```
Comment (Godofnothing) [inline src/sparseml/pytorch/sparsification/pruning/modifier_pruning_obs.py:243]: personally, I think giving **kwargs would be more convenient in some cases
